packages/@lwc/engine-core/src/libs/signal/index.ts

```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright (c) 2023, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+
+export type OnUpdate = () => void;
+export type Unsubscribe = () => void;
+
+export interface Signal<ValueShape> {
+    get value(): ValueShape;
+    subscribe(onUpdate: OnUpdate): Unsubscribe;
+}
+
+export abstract class SignalBaseClass<ValueShape> implements Signal<ValueShape> {
+    abstract get value(): ValueShape;
+
+    private subscribers: Set<OnUpdate> = new Set();
+
+    subscribe(onUpdate: OnUpdate) {
+        this.subscribers.add(onUpdate);
+        return () => {
+            this.subscribers.delete(onUpdate);
+        };
+    }
+
+    protected notify() {
+        for (const subscriber of this.subscribers) {
+            subscriber();
+        }
+    }
+}
```

Comment:
As we discussed, I'd love for this to be moved to a new top-level `@lwc/signals` package that exposes the types and base class. I would have recommended moving to `@lwc/shared`, except that is intended as an internal package with other `@lwc/*` packages being the only direct consumers. In contrast, the signals type and base call may be re-used across the SF ecosystem.

Reply:
@divmain I moved everything to a new [package](https://github.com/salesforce/lwc/blob/7b3582faaf6bd3365091d24740706d46cd8353d9/packages/%40lwc/signals/README.md) and added a `readme`. 

If you get a chance, can you look through the `readme` too?